### ISSUE TITLE
robot-simulator requires global state

### ISSUE THREAD

**Issue by katsel** [exercism/xerlang#123]
I tried to do the robot-simulator exercise today and found it very difficult. It seems like one has to use global state for the example - functions to manipulate the robot [don't have a return value](https://github.com/exercism/xerlang/blob/master/exercises/robot-simulator/robot_simulator_tests.erl#L12), so the current state of the robot needs to be kept somewhere else. I found that very confusing.

As far as I could see, people doing this exercise have come up with 3 different solutions for this:
* a) use a FSM with either message passing/receiving or the `gen_fsm` pattern
* b) store state in an `ets` table
* c) modify the tests to return a new robot each time there is a change in state.

a) is rather complicated and requires advanced knowledge of Erlang-specific patterns that I wouldn't assume most beginners have. Even if beginners manage to find out about these patterns, all these things distracts from the actual exercise (which is difficult enough on its own). 

b) is more simple to implement (once you know what you are looking for), but it's actively discouraging people from using functional programming style like you'd usually do as an Erlang beginner, and as the other exercises have shown.

c) Seems to be the most natural way to do this in Erlang - generate a robot, and each time you manipulate it, return a new robot object with a new, appropriate state and continue working with that object. Currently that'd be a hack.

Long story short, I'd suggest to change the exercise so one doesn't need to worry about global state and/or fsm patterns to solve it. Making the robot behave as expected is challenging enough on its own - worries about 'where do I store the state of my robot' should not be such a large part of the exercise.

**Comment by NobbZ** [comment]
I haven't seen a `gen_fsm` solution so far, but many that roll out an own process or use `gen_server`.

And actually this exercise is meant to draw attention to a state-holding process. It is planned though to add some HINTS.md during the huge overhaul.

**Comment by katsel** [comment]
Could one reconsider the placement of this exercise then? The preceeding and following examples are *way* easier and require *way* less effort (both in time needed to solve and Erlang-specific knowledge).

Also, it might be a good idea to start off with a trivial exercise on state-holding, to get people used to this pattern in a way that's less frustrating. Having to solve a non-trivial roboter-programming exercise at the same time is just tough.

**Comment by NobbZ** [comment]
I'm very open to reordering and collecting input. Sadly all input I have so far is my own. Depending on the amount of input I do plan to reorder some when during December.

**Comment by NobbZ** [comment]
I forgot to metion it before. Re-ordering the exercises is part of #107.

**Comment by katsel** [comment]
I'd like to help with the reordering! 

I looked at the xerlang-reclassification repo before, but to be hondest, I couldn't make any sense of the difficulty-measurements. There seems to be no scale. Does it range from 1-10? Who is the "yardstick" for this - beginner or Erlang pro? Programming beginner or someone with prior experience in other languages? What actually is "difficulty" - the time it took to complete the exercise, the number of lines typed, the amount of math knowledge required?
Considering the many features involved here, I just feel I cannot put "difficulty" into a simple number between 1 and 10.
That's why I shied back from participating in your repo.

My suggestion would be: Keep the ordering of most exercises and only take a closer look at the "outliers" - exercises that are clearly more difficult than those before and after them, and try to find a new placement for those.
Simply looking at the number of people that submitted a valid solution in exercism already is helpful gauging that, and of course asking the learners in the track about their experience.

Also, mind that many exercises could be made much easier and approachable by just adding a small note in the README, telling the learners how to tackle the problem. I think that would be a useful thing to do before reordering everything. And I'd gladly be helping with that, too.

**Comment by NobbZ** [comment]
If the difficulty were easily measurable in an objective way I had done it already and sorted accordingly, but a difficulty is a very subjective thing.

Eg. I do program (as a hobby and self taught) since about 1990, I started studying CS 3 years ago. I just used processes to solve the robot exercise without even thinking about it. Others do find this aspect of the exercise very hard to come around as I do understand your issue.

But yes, the range you guessed is correct, it is from 1 (easiest) to 10 (hardest). This correlates to the specification of the JSON.

Another thing is, that I do not have any possibility to alter the README of an exercise. It is assembled from a central README for that exercise, roughly describing the task as well as a generic part for the language which should explain how to run a test, but this part is specific to the hole track.

It is common though through the tracks to have an HINTS.md or hints folder, which contains additional information or stuff to help out with the exercise. Adding those is on my TODO.

But I have to admit, my 2016 overhaul will probably not be finished before February or March 2017 :(

**Comment by katsel** [comment]
Thanks for that explanation! I'll see what I can do. 😊
